ggml-cuda.cu

```diff
@@ -726,115 +645,335 @@ static void ggml_cuda_mul_mat_f16(const ggml_tensor * src0, const ggml_tensor *
+                cudaStream_t cudaStream_memcpy_src1 = g_cudaStreams_memcpy_src1[id][i0 % GGML_CUDA_MAX_STREAMS];
+                cudaStream_t cudaStream_memcpy_dst = g_cudaStreams_memcpy_dst[id][i0 % GGML_CUDA_MAX_STREAMS];
+                cudaEvent_t  cudaEvent_main = g_cudaEvents_main[id][i0 % GGML_CUDA_MAX_EVENTS];
+                cudaEvent_t  cudaEvent_memcpy_src1 = g_cudaEvents_memcpy_src1[id][i0 % GGML_CUDA_MAX_EVENTS];
+
+                char  * src0_ddq_i = src0_ddq[id] + (i0 - row_low/ne01)*src0_stride;
+                float * src0_ddf_i = src0_ddf[id] + (i0 - row_low/ne01)*src0_stride;
+                float * src1_ddf_i = src1_ddf[id] + i1*src1_stride;
+                float * dst_ddf_i = dst_ddf[id] + (i0 - row_low/ne0)*dst_stride;
+
+                // copy src0, src1 to device if necessary
+                if (!src1_on_device) {
+                    CUDA_CHECK(ggml_cuda_h2d_tensor_2d(src1_ddf_i, src1, i03, i02, 0, ne11, cudaStream_memcpy_src1));
+                }
+                CUDA_CHECK(cudaEventRecord(cudaEvent_memcpy_src1, cudaStream_memcpy_src1));
+                if (!src0_on_device) {
+                    if (src0_is_f32) {
+                        CUDA_CHECK(ggml_cuda_h2d_tensor_2d(src0_ddf_i, src0, i03, i02, i01_low, i01_high, cudaStream_main));
+                    } else {
+                        CUDA_CHECK(ggml_cuda_h2d_tensor_2d(src0_ddq_i, src0, i03, i02, i01_low, i01_high, cudaStream_main));
+                    }
+                }
 
-                // wait for conversion
-                CUDA_CHECK(cudaStreamWaitEvent(cudaStream, cudaEvent, 0));
+                if (src0_needs_f32 && !src0_is_f32) {
+                    to_fp32_cuda(
+                        src0_ddq_i,
+                        src0_ddf_i,
+                        i01_diff*ne00, cudaStream_main);
+                    CUDA_CHECK(cudaGetLastError());
+                }
 
-                // compute
-                CUBLAS_CHECK(cublasSetStream(g_cublasH, cudaStream));
-                CUBLAS_CHECK(
-                    cublasSgemm(g_cublasH, CUBLAS_OP_T, CUBLAS_OP_N,
-                            ne01, ne11, ne10,
-                            &alpha, c_X, ne00,
-                                    c_Y, ne10,
-                            &beta,  c_D, ne01));
+                // wait with main stream until src1 memcpy is done
+                CUDA_CHECK(cudaStreamWaitEvent(cudaStream_main, cudaEvent_memcpy_src1, 0));
+
+                // do the computation
+                op(src0, src1, dst, src0_ddq_i, src0_ddf_i, src1_ddf_i, dst_ddf_i, i01_low, i01_high, i1, cudaStream_main);
+
+                CUDA_CHECK(cudaEventRecord(cudaEvent_main, cudaStream_main));
+
+                // copy dst to host if necessary
+                if (!dst_on_device) {
+                    // wait with memcpy until main stream is done
+                    CUDA_CHECK(cudaStreamWaitEvent(cudaStream_memcpy_dst, cudaEvent_main, 0));
```

Comment:
Why not do this in the stream `cudaStream_main` instead? Is there any parallelization gained by doing this?

Reply:
There does not seem to be a measurable performance difference. I added these because I think they will be useful for synchronization: you can wait for computation to finish by synchronizing the main streams and for data transfer to finish by waiting for the data copy streams. It wouldn't be a problem to remove them until they are actually useful though.

Reply:
If it is still not useful, I think it is better to remove it just to reduce the complexity of the code. I am not sure in what situations it may be desirable to wait for the computation instead of the copy, if you plan to do further computations on the GPU then the copy is not needed anyway.